src/package-resolver.js

```diff
@@ -530,10 +530,48 @@ export default class PackageResolver {
       this.resolveToResolution(req);
     }
 
+    deps.forEach(dep => {
+      const name = normalizePattern(dep.pattern).name;
+      this.optimizeResolutions(name);
+    });
+
     activity.end();
     this.activity = null;
   }
 
+  // for a given package, see if a single manifest can satisfy all ranges
+  optimizeResolutions(name: string) {
+    const patterns: Array<string> = this.patternsByPackage[name] || [];
+
+    // don't optimize things that already have a lockfile entry:
+    // https://github.com/yarnpkg/yarn/issues/79
+    const anyLocked = patterns.some(pattern => this.lockfile.getLocked(pattern));
+    if (anyLocked) {
+      return;
```

Comment:
Why not exclude only the locked ones and still optimize the rest?

Reply:
So I've made this change, but it causes some workspace related tests to fail ("install should link workspaces that refer each other", "install should not link workspaces that refer not compatible version of another workspace"). I don't quite understand why these are failing, but I've just disabled the optimization wherever `remote.type === "workspace"`. Open to other suggestions.